### PR TITLE
WIP: Gherkin BDD for e2e support via godog

### PR DESCRIPTION
**Pull request by hh**
**What type of PR is this?**

/kind design

**What this PR does / why we need it**:

```feature
Feature: Support e2e tests via new godog BDD framework with separate behavior definitions
  As a time constrained behavior or test reviewer
  I want to be able to review tests separately from behaviors
    And maintain backwards compatibility with the existing e2e framework
  In order to speed of the creation of behaviors and tests I can review
  Scenario:
    Given an existing e2e test
    When I refer to it within the new gherkin/go-dog framework
    Then the behavior metadata will connect the old test to the new framework
```

**Which issue(s) this PR fixes**:
Implements initial infra for [BDD Conformance Testing KEP](https://github.com/kubernetes/enhancements/pull/959)

**Special notes for your reviewer**:

Currently evolving, but with this PR you can run e2e.test --godog and it will pick up features from `test/e2e/features` and step definitions from `test/e2e/features/steps`.

```shell
bazel build //test/e2e:e2e.test && KUBECONFIG=$(kind  get kubeconfig-path) ./bazel-bin/test/e2e/e2e.test --godog
WARNING: Ignoring JAVA_HOME, because it must point to a JDK, not a JRE.
INFO: Analyzed target //test/e2e:e2e.test (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //test/e2e:e2e.test up-to-date:
  bazel-bin/test/e2e/e2e.test
INFO: Elapsed time: 0.556s, Critical Path: 0.20s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
Jul  9 13:52:25.671: INFO: The --provider flag is not set. Continuing as if --provider=skeleton had been used.
```

```feature
Feature: Use existing ginkgo framework
  As a test contributor
  I want to not throw away all our old tests
  In order to retain the value generated in them

  Scenario: Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly                                                        # test/e2e/features/ginkgo/existing_tests.feature:6
    Given existing test "[k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly [NodeConformance] [Conformance]" # first_steps.go:8 -> k8s.io/kubernetes/.existingTest
      TODO: write pending definition
    When I run the test                                                                                                                                                       # first_steps.go:12 -> k8s.io/kubernetes/.iRunTheTest
    Then the existing test will pass                                                                                                                                          # first_steps.go:16 -> k8s.io/kubernetes/.theExistingTestWillPass
    And this is fine                                                                                                                                                          # first_steps.go:20 -> k8s.io/kubernetes/.thisIsFine
```

```
1 scenarios (1 pending)
4 steps (1 pending, 3 skipped)
205.804µs

Randomized with seed: 1562637145671052559
PASS
```

**Does this PR introduce a user-facing change?**:

```release-note
Add support for Gherkin based e2e tests using godog
```

### CODE REVIEW

File: test/e2e/features/ginkgo/existing_tests.feature [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Feature: Use existing ginkgo framework
+  As a test contributor
+  I want to not throw away all our old tests
+  In order to retain the value generated in them
+  @ginko @Conformance @NodeConformance @k8s.io
+  Scenario: Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly
+    Given existing test "[k8s.io] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly [NodeConformance] [Conformance]"
```
Comment (hh) [inline test/e2e/features/ginkgo/existing_tests.feature]: This is usually run with ```e2e.test --ginkgo.focus "\[k8s.io\] Container Lifecycle Hook when create a pod with lifecycle hook should execute poststart exec hook properly \[NodeConformance\] \[Conformance\]"```
I'm hoping we can find a way to run the test directly, sans before and aftersuite calls.

File: test/e2e/features/steps/first_steps.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	// BeforeScenario(func(interface{}))
+	// BeforeSuite(func())
+	// AfterSuite(func())
+	// AfterScenario(func(interface{}))
+	// AfterScenario(func(interface{}))
+}
+
+var testingMain *testing.M
+
+func existingTest(existingTestName string) error {
```
Comment (hh) [inline test/e2e/features/steps/first_steps.go]: This method should be able to take the existing test name and call the Ginkgo Framework via m.Run() or similar.

I'm having hard time finding the right entry point.

**Comment by timothysc** [comment]
So I'm not sold that gherkin would be any better.  In conversing last time I wanted to start from a requirements perspective first.  For example, let's write an ideal end state then try to figure out a way to incrementally get there.

**Pull request closed without merging**